File: lib/expression-parser.ts
/**
 * Safe expression evaluator.
 * Pure manual tokenization and interpretation.
 * Uses NO dynamic code execution of any kind.
 */

/**
 * Split a string on a delimiter, but only at the top level (ignoring splits inside parentheses).
 */
function splitOutsideParens(str: string, delimiter: string): string[] {
  const parts: string[] = []
  let depth = 0
  let current = ""
  let i = 0

  while (i < str.length) {
    if (str[i] === "(") {
      depth++
      current += str[i]
      i++
    } else if (str[i] === ")") {
      depth--
      current += str[i]
      i++
    } else if (depth === 0 && str.startsWith(delimiter, i)) {
      parts.push(current)
      current = ""
      i += delimiter.length
    } else {
      current += str[i]
      i++
    }
  }
  parts.push(current)
  return parts
}

/**
 * Resolve a value token against context.
 * Handles: string literals ("..." or '...'), number literals, true/false/null, dot-path lookups.
 */
function resolveValue(token: string, context: Record<string, any>): any {
  const t = token.trim()

  // String literal (double or single quotes)
  if ((t.startsWith('"') && t.endsWith('"')) || (t.startsWith("'") && t.endsWith("'"))) {
    return t.slice(1, -1)
  }

  // Boolean literals
  if (t === "true") return true
  if (t === "false") return false

  // Null
  if (t === "null") return null

  // Number literal
  if (/^-?\d+(\.\d+)?$/.test(t)) return Number(t)

  // Context path (dot notation)
  const parts = t.split(".")
  let current: any = context
  for (const part of parts) {
    if (current === undefined || current === null) return undefined
    current = current[part]
  }
  return current
}

/**
 * Evaluate an atomic comparison expression (no logical operators at top level).
 * Handles: ==, !=, >=, <=, >, <
 * Also handles bare truthy expressions (no operator).
 */
function evaluateComparison(expr: string, context: Record<string, any>): boolean {
  const trimmed = expr.trim()

  // Check for comparison operators (ordered: multi-char before single-char)
  const operators = [">=", "<=", "!=", "==", ">", "<"]
  for (const op of operators) {
    const idx = trimmed.indexOf(op)
    if (idx !== -1) {
      const left = trimmed.slice(0, idx)
      const right = trimmed.slice(idx + op.length)
      const lv = resolveValue(left, context)
      const rv = resolveValue(right, context)

      switch (op) {
        case "==": return lv == rv // intentional loose equality for null/"" cases
        case "!=": return lv != rv
        case ">=": return Number(lv) >= Number(rv)
        case "<=": return Number(lv) <= Number(rv)
        case ">":  return Number(lv) > Number(rv)
        case "<":  return Number(lv) < Number(rv)
      }
    }
  }

  // Bare truthy: just a variable name or path
  return Boolean(resolveValue(trimmed, context))
}

/**
 * Evaluate a unit (handles parentheses and NOT prefix).
 */
function evaluateUnit(expr: string, context: Record<string, any>): boolean {
  const trimmed = expr.trim()

  // NOT prefix
  if (trimmed.startsWith("!")) {
    return !evaluateUnit(trimmed.slice(1), context)
  }

  // Parentheses
  if (trimmed.startsWith("(") && trimmed.endsWith(")")) {
    return evaluateExpression(trimmed.slice(1, -1), context)
  }

  return evaluateComparison(trimmed, context)
}

/**
 * Evaluate an expression string with full support for:
 * - Comparison: ==, !=, >, <, >=, <=
 * - Logical AND (&&), OR (||)
 * - NOT (!)
 * - Parentheses
 * - Nested context paths (dot notation)
 * - Bare truthy values
 *
 * SECURITY: Pure manual parsing only. No dynamic code execution.
 */
export function evaluateExpression(expression: string, context: Record<string, any>): boolean {
  const trimmed = expression.trim()
  if (!trimmed) return false

  // Split on || (lowest precedence) — outside parens
  const orParts = splitOutsideParens(trimmed, "||")
  if (orParts.length > 1) {
    return orParts.some((part) => evaluateExpression(part.trim(), context))
  }

  // Split on && — outside parens
  const andParts = splitOutsideParens(trimmed, "&&")
  if (andParts.length > 1) {
    return andParts.every((part) => evaluateExpression(part.trim(), context))
  }

  // Single unit
  return evaluateUnit(trimmed, context)
}

File: sdk/python/composr/expression_parser.py
"""
Safe expression evaluator — pure manual parsing.
Supports: ==, !=, >, <, >=, <=, &&, ||, !, parentheses, dot-path context lookups.
"""
from __future__ import annotations
from typing import Any


def _split_outside_parens(s: str, delimiter: str) -> list[str]:
    parts: list[str] = []
    depth = 0
    current = ""
    i = 0
    while i < len(s):
        if s[i] == "(":
            depth += 1
            current += s[i]
            i += 1
        elif s[i] == ")":
            depth -= 1
            current += s[i]
            i += 1
        elif depth == 0 and s[i:i + len(delimiter)] == delimiter:
            parts.append(current)
            current = ""
            i += len(delimiter)
        else:
            current += s[i]
            i += 1
    parts.append(current)
    return parts


def _resolve_value(token: str, context: dict[str, Any]) -> Any:
    t = token.strip()
    if (t.startswith('"') and t.endswith('"')) or (t.startswith("'") and t.endswith("'")):
        return t[1:-1]
    if t == "true":
        return True
    if t == "false":
        return False
    if t == "null":
        return None
    try:
        return int(t) if "." not in t else float(t)
    except ValueError:
        pass
    # Dot-path lookup
    current: Any = context
    for part in t.split("."):
        if current is None or not isinstance(current, dict):
            return None
        current = current.get(part)
    return current


def _evaluate_comparison(expr: str, context: dict[str, Any]) -> bool:
    trimmed = expr.strip()
    for op in (">=", "<=", "!=", "==", ">", "<"):
        idx = trimmed.find(op)
        if idx != -1:
            left = _resolve_value(trimmed[:idx], context)
            right = _resolve_value(trimmed[idx + len(op):], context)
            if op == "==":
                return left == right
            if op == "!=":
                return left != right
            try:
                lv, rv = float(left), float(right)
            except (TypeError, ValueError):
                return False
            if op == ">=":
                return lv >= rv
            if op == "<=":
                return lv <= rv
            if op == ">":
                return lv > rv
            if op == "<":
                return lv < rv
    return bool(_resolve_value(trimmed, context))


def _evaluate_unit(expr: str, context: dict[str, Any]) -> bool:
    trimmed = expr.strip()
    if trimmed.startswith("!"):
        return not _evaluate_unit(trimmed[1:], context)
    if trimmed.startswith("(") and trimmed.endswith(")"):
        return evaluate_expression(trimmed[1:-1], context)
    return _evaluate_comparison(trimmed, context)


def evaluate_expression(expression: str, context: dict[str, Any]) -> bool:
    trimmed = expression.strip()
    if not trimmed:
        return False
    or_parts = _split_outside_parens(trimmed, "||")
    if len(or_parts) > 1:
        return any(evaluate_expression(p.strip(), context) for p in or_parts)
    and_parts = _split_outside_parens(trimmed, "&&")
    if len(and_parts) > 1:
        return all(evaluate_expression(p.strip(), context) for p in and_parts)
    return _evaluate_unit(trimmed, context)
